[jin.100] 방탈출 예약 대기 관리 Step 1 & 2

dongjinBaek opened the pull request:
제프리 안녕하세요.
방탈출 예약 대기 관리 1,2단계 코드 작성했습니다.

# 1단계 : 인증 로직 리팩터링

## 구현

- 인증 로직에 사용되던 `nextstep.auth` 패키지를 `nextstep` 패키지 밖으로 이동시켰습니다.
- `auth` 패키지가 `nextstep` 패키지에 의존하지 않도록 변경하는 것이 1단계 목표였습니다.
- 이동 직후 `auth` 패키지에서 `Member`, `MemberDao` 클래스에 의존하고 있어서 각각에 해당하는  각각이 하던 일을 담당하는 `UserDetails`, `UserDetailsService`를 만들었습니다.
- `nextstep.member` 패키지에서 `UserDetails`를 상속받은 `Member` 클래스, `UserDetailsService`를 구현한 `LoginMemberService`를 만들었습니다.
- 만들고 나서 `UserDetailsService`에 `findById(Long id)` 메서드가 있어서 근데 이러면 몽고디비 쓰고 싶은 사람은 이 패키지 못 쓰는거 아닌가? 싶어서 제네릭으로 변경했습니다.

## 고민
- `auth` 패키지를 사용하려면 `UserDetails`타입 클래스, `UserDetailsService` 타입 빈을 하나 이상 정의해야 하는 구조인데요, 이런 내용들을 코드에 어떻게 잘 담을 수 있을까요? 메모 잘 남겨두기 외에 좋은 방법이 떠오르지 않아서 고민이었습니다.

# 2단계 : 예약 대기 기능

## 구현
- 이미 예약된 스케줄에 대해 예약 대기를 걸 수 있는 기능을 추가했습니다.
- 예약 대기 신청 시 해당 스케줄에 예약이 없으면 바로 예약이 되는 요구사항이 있었습니다.
- 예약 생성 시 `Reservation`과 `Waiting` 객체를 멤버로 가지는 (둘 중 하나는 null) `CreateWaitingResponse` 타입을 서비스에서 리턴하고 컨트롤러에서 이를 바탕으로 헤더 Location 지정하도록 했습니다.
- 예약 대기 번호를 알려주기 위해서 `WaitingDao`에서 예약 대기 조회 시 SELECT 절에 subquery로 해당 스케줄에 있는 대기자 수도 쿼리하도록 했습니다.

## 고민
- 지금은 `CreateWaitingResponse`에 필드들을 `final` 로 만들고, 정적 팩토리 메서드를 통해 둘 중 하나가 null이라는 조건을 만족시키고 있습니다. 어떻게 하면 나중에 여러 사람들이 이 클래스를 수정하더라도(팩토리 메서드 추가 등) 둘 중 하나가 null 이라는 제약사항이 지켜지도록 할 수 있을까요?
- 예약 대기 번호를 subquery로 호출하는게 좋은 방법인지?

kakao-jin-100 committed:
docs : 1단계 요구사항 정리

kakao-jin-100 committed:
refactor : nextstep.auth 패키지 auth 로 변경

kakao-jin-100 committed:
refactor : UserDetails, UserDetailsService 추가

kakao-jin-100 committed:
docs : 2단계 요구사항 추가

kakao-jin-100 committed:
refactor : UserDetailsService 제네릭으로 변경

kakao-jin-100 committed:
feat : 예약대기 생성기능 추가

kakao-jin-100 committed:
feat : 토큰이 없을 때 @LoginMember에 null 대신 AuthenticationException 던지도록 변경

kakao-jin-100 committed:
feat : 내 예약 목록 조회 기능 추가

kakao-jin-100 committed:
feat : 내 예약 대기 목록 조회 기능 추가

kakao-jin-100 committed:
refactor : 내 예약 조회 응답 member 정보 제거

kakao-jin-100 committed:
refactor : WaitingController location prefix 상수로 분리

kakao-jin-100 committed:
refactor : WaitingRequest 를 CreateWaitingRequest 로 rename

kakao-jin-100 committed:
feat : 다른 사람의 예약을 삭제하는 경우 AuthenticationException 대신 ForbiddenException 을 던지도록 변경

kakao-jin-100 committed:
feat : 예약 대기 삭제 기능 추가

kakao-jin-100 committed:
fix : 예약 대기번호 조회 시 나보다 일찍 예약 대기한 사람의 수만 세도록 변경

artium59 commented:
> 만들고 나서 UserDetailsService에 findById(Long id) 메서드가 있어서 근데 이러면 몽고디비 쓰고 싶은 사람은 이 패키지 못 쓰는거 아닌가? 싶어서 제네릭으로 변경했습니다.

`UserDetails`을 제네릭으로 구현하셔서 `UserDetails`타입 빈과 `UserDetailsService`타입 서비스를 하나 이상 구현하는 것 같은데요, 굳이 제네릭으로 구현할 필요가 있나 생각이 되네요.

`UserDetailsService`가 단 하나의 class/interface를 반환하게 바꿔보시겠어요?
저장소가 몽고DB로 바뀐다고 해도 사용은 가능합니다. 새로운 `LoginMemberService`를 만들기만 하면 되니까요.

혹시 더 자세한 설명이 필요하시다면 말씀해주세요.

p.s. 스프링의 UserDetails는 일반 인터페이스입니다.
<br>
> 어떻게 하면 나중에 여러 사람들이 이 클래스를 수정하더라도(팩토리 메서드 추가 등) 둘 중 하나가 null 이라는 제약사항이 지켜지도록 할 수 있을까요?

아니요. 힘들다고 생각합니다. 온갖 제한을 걸어도(ex. 둘 다 null이 아님 exception을 던진다던가), 코드 수정이 가능한 순간 제약은 사라진다고 생각합니다.
저라면 `CreateWaitingResponse`의 파라미터를 isReservedDirectly와 id만 갖고 있을 것 같네요.

> 예약 대기 번호를 subquery로 호출하는게 좋은 방법인지?

저 쿼리는 where 절의 id가 유동적이라서 어떻게 안되겠네요.

artium59 commented on src/main/java/nextstep/member/Member.java:
```diff
@@ -1,13 +1,8 @@
-    private String role;
+import auth.UserDetails;
 
+public class Member extends UserDetails<Long> {
```
멤버 클래스는 엔티티인데, `UserDetails`를 상속받네요. `UserDetails`가 변할 때마다 DB도 바뀌겠네요.

artium59 commented on src/main/java/auth/AuthConfiguration.java:
```diff
@@ -0,0 +1,43 @@
+package auth;
```
패키지를 간단하게라도 분리해주셨으면 합니다.
exception, configure, etc...

artium59 reviewed (changes requested):
진 안녕하세요.
간단하게 리뷰 남겼습니다. 확인 부탁드려요.

kakao-jin-100 committed:
chore : auth 패키지 내 세부 패키지 분리

kakao-jin-100 committed:
Revert "refactor : UserDetailsService 제네릭으로 변경"

This reverts commit eeb8674ae77f3d7d9738b03d8014713d3d2fbea6.

kakao-jin-100 committed:
refactor : 로그인한 유저 정보 가져오는 어노테이션 LoginUser 로 rename

kakao-jin-100 committed:
feat : UserDetails를 상속받은 LoginMember 클래스 추가

- 기존에는 Entity인 Member 클래스가 UserDetails를 상속받았는데, 새로운 클래스를 만들었습니다

kakao-jin-100 committed:
refactor : CreateWaitingResponse 내부 멤버 변수들 변경

- 기존에 reservation, waiting 객체를 멤버로 가지고, getter 함수에서 계산해서 리턴했었음
- 변경 이후 조회하는 isReservedDirectly와 id 값을 멤버 변수로 가지고, 정적 팩토리 메서드에서 계산해서 넣어주도록 함

dongjinBaek commented:
제프리 안녕하세요.
리뷰 남겨주신 내용 반영해서 수정했습니다.

찾아보니까 스프링 `UserDetails`에는 id 값 없이 username, password만 가지고 있었네요.

artium59 reviewed (approved):
고생하셨습니다. 머지하도록 할게요:)

artium59 commented:
> 찾아보니까 스프링 UserDetails에는 id 값 없이 username, password만 가지고 있었네요.

`UserDetails`의 [getAuthorities()](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/core/userdetails/UserDetails.html#getAuthorities())는 `getRole()`과 비슷하다고 생각하시면 됩니다.
